sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/errorhandling/ErrorHandler.java

```diff
@@ -0,0 +1,187 @@
+
+    private boolean closed = false;
+
+    /**
+     * Constructs a new ErrorHandler, but should not be called directly. Instead, call
+     * pipeline.registerErrorHandler to ensure safe pipeline construction
+     */
+    @Internal
+    public PTransformErrorHandler(PTransform<PCollection<ErrorT>, OutputT> sinkTransform) {
+      this.sinkTransform = sinkTransform;
+    }
+
+    @Override
+    public void addErrorCollection(PCollection<ErrorT> errorCollection) {
+      errorCollections.add(errorCollection);
+    }
+
+    @Override
+    public boolean isClosed() {
+      return closed;
+    }
+
+    @Override
+    public OutputT getOutput() {
+      if (!this.isClosed()) {
+        throw new IllegalStateException(
+            "ErrorHandler must be finalized before the output can be returned");
+      }
+      // make the static analysis checker happy
+      Preconditions.checkArgumentNotNull(sinkOutput);
+      return sinkOutput;
+    }
+
+    @Override
+    public void close() {
+      closed = true;
+      if (errorCollections.isEmpty()) {
+        LOG.warn("Empty list of error pcollections passed to ErrorHandler.");
+        return;
+      }
+      LOG.debug(
+          "{} error collections are being sent to {}",
+          errorCollections.size(),
+          sinkTransform.getName());
+      sinkOutput =
+          PCollectionList.of(errorCollections)
+              .apply(Flatten.pCollections())
+              .apply(new WriteErrorMetrics<ErrorT>(sinkTransform))
+              .apply(
+                  sinkTransform.addAnnotation(
+                      "FeatureMetric", "ErrorHandler".getBytes(StandardCharsets.UTF_8)));
+    }
+
+    public static class WriteErrorMetrics<ErrorT>
+        extends PTransform<PCollection<ErrorT>, PCollection<ErrorT>> {
+
+      private final Counter errorCounter;
+
+      public WriteErrorMetrics(PTransform<?, ?> sinkTransform) {
+        errorCounter = Metrics.counter("ErrorMetrics", sinkTransform.getName() + "-input");
+      }
+
+      @Override
+      public PCollection<ErrorT> expand(PCollection<ErrorT> input) {
+        return input.apply(ParDo.of(new CountErrors<ErrorT>(errorCounter)));
+      }
+
+      public static class CountErrors<ErrorT> extends DoFn<ErrorT, ErrorT> {
+
+        private final Counter errorCounter;
+
+        public CountErrors(Counter errorCounter) {
+          this.errorCounter = errorCounter;
+        }
+
+        @ProcessElement
+        public void processElement(@Element ErrorT error, OutputReceiver<ErrorT> receiver) {
+          errorCounter.inc();
+          receiver.output(error);
+        }
+      }
+    }
+  }
+
+  @Internal
+  class NoOpErrorHandler<ErrorT, OutputT extends POutput> implements ErrorHandler<ErrorT, OutputT> {
+
+    @Override
+    public void addErrorCollection(PCollection<ErrorT> errorCollection) {}
+
+    @Override
+    public boolean isClosed() {
+      throw new IllegalArgumentException("No Op handler should not be closed");
```

Comment:
Perhaps we should still allow closing this one? (It's unclear what the intended use is. If it's to ignore errors, let's make it a bit scarier sounding.)

Reply:
I've added some comments. Basically, the only method that should get called when this is used is addErrorCollection, and the other methods imply misconfiguration